### DISCUSSION TITLE
Permutation Sensitive Aggregations?

### DISCUSSION THREAD

**Discussion by jackeown** [pyg-team/pytorch_geometric#2110] [Q&A]
I was reading [this great tutorial](http://pages.di.unipi.it/citraro/files/slides/Landolfi_tutorial.pdf) and I noticed a few things.

1. There is a `torch.scatter` function. This is quite different from PyG scatter functions, right?
2. On page 18, they say that in order to use the `torch_geometric.nn.MessagePassing`, you need to have a permutation invariant `aggregate` function. I understand that this is the common use-case, but is it absolutely necessary?

The reason I ask for a permutation sensitive aggregation is that I'm working on a dataset of DAGs for which the order of a node's children is quite important.

**Comment by rusty1s** [comment]
Thank your for the link to this great tutorial. I added it to our "External Resources" page:
1. The `torch.scatter` function will select a random value from a set of values with same index. PyG scatter functions will reduce values from all indices instead. Recently, there is some effort to replicate this behaviour directly in PyTorch (`torch.scatter` now supports a `reduce` argument), but same reductions are net yet supported (such as `max`).
2. That is correct. Currently, we only support permutation-invariant aggregations defined in `torch-scatter`. To use other aggregations, you may need to pad your neighboring features into shape `[num_nodes, num_neighbors, num_features]`, but currently, you are required to do this by yourself.
